File: kits/data-integration/providers/quality_rule/reconciliation.ts
// Quality Rule Provider: Reconciliation Validation
// Matches field values against external knowledge bases for accuracy verification.
// Dimension: accuracy

export const PROVIDER_ID = 'reconciliation';
export const PLUGIN_TYPE = 'quality_rule';

export interface FieldDef {
  name: string;
  type: string;
  required?: boolean;
  constraints?: Record<string, unknown>;
}

export interface RuleConfig {
  options?: Record<string, unknown>;
  threshold?: number;
}

export interface RuleResult {
  valid: boolean;
  message?: string;
  severity: 'error' | 'warning' | 'info';
}

export type QualityDimension = 'completeness' | 'uniqueness' | 'validity' | 'consistency' | 'timeliness' | 'accuracy';

interface KBMatch {
  canonicalValue: string;
  confidence: number;
  source: string;
}

type KBQueryFn = (value: string, config: Record<string, unknown>) => Promise<KBMatch[]>;

export class ReconciliationQualityProvider {
  private knowledgeBaseHandlers: Map<string, KBQueryFn> = new Map();

  constructor() {
    this.registerDefaultHandlers();
  }

  /**
   * Register a custom knowledge base query handler.
   */
  registerKnowledgeBase(name: string, handler: KBQueryFn): void {
    this.knowledgeBaseHandlers.set(name, handler);
  }

  async validateAsync(
    value: unknown,
    field: FieldDef,
    _record: Record<string, unknown>,
    config: RuleConfig
  ): Promise<RuleResult> {
    if (value === null || value === undefined) {
      return { valid: true, severity: 'info' };
    }

    const knowledgeBase = (config.options?.knowledgeBase as string) ?? 'wikidata';
    const matchThreshold = config.threshold ?? 0.8;
    const stringValue = String(value);

    const handler = this.knowledgeBaseHandlers.get(knowledgeBase);
    if (!handler) {
      return {
        valid: false,
        message: `Reconciliation rule for '${field.name}': unknown knowledge base '${knowledgeBase}'.`,
        severity: 'warning',
      };
    }

    try {
      const matches = await handler(stringValue, config.options ?? {});

      if (matches.length === 0) {
        return {
          valid: false,
          message: `Field '${field.name}' value '${stringValue}' not found in knowledge base '${knowledgeBase}'.`,
          severity: 'warning',
        };
      }

      const bestMatch = matches.reduce((best, m) =>
        m.confidence > best.confidence ? m : best
      );

      if (bestMatch.confidence >= matchThreshold) {
        if (bestMatch.canonicalValue !== stringValue) {
          return {
            valid: true,
            message: `Field '${field.name}': canonical form is '${bestMatch.canonicalValue}' (confidence: ${(bestMatch.confidence * 100).toFixed(1)}%, source: ${bestMatch.source}).`,
            severity: 'info',
          };
        }
        return { valid: true, severity: 'info' };
      }

      return {
        valid: false,
        message: `Field '${field.name}' value '${stringValue}' has low confidence match (${(bestMatch.confidence * 100).toFixed(1)}%) against '${knowledgeBase}'. Best match: '${bestMatch.canonicalValue}'.`,
        severity: 'warning',
      };
    } catch (err) {
      return {
        valid: false,
        message: `Reconciliation error for '${field.name}': ${(err as Error).message}`,
        severity: 'warning',
      };
    }
  }

  /**
   * Synchronous validate for interface compatibility.
   * For real KB lookups, use validateAsync instead.
   */
  validate(
    value: unknown,
    field: FieldDef,
    _record: Record<string, unknown>,
    config: RuleConfig
  ): RuleResult {
    if (value === null || value === undefined) {
      return { valid: true, severity: 'info' };
    }

    const stringValue = String(value);
    const matchThreshold = config.threshold ?? 0.8;

    // Synchronous local-dictionary check when no async KB is needed
    const localValues = config.options?.localDictionary as string[] | undefined;
    if (localValues && localValues.length > 0) {
      const bestMatch = this.findBestLocalMatch(stringValue, localValues);
      if (bestMatch.confidence >= matchThreshold) {
        if (bestMatch.canonicalValue !== stringValue) {
          return {
            valid: true,
            message: `Field '${field.name}': suggested canonical form is '${bestMatch.canonicalValue}' (similarity: ${(bestMatch.confidence * 100).toFixed(1)}%).`,
            severity: 'info',
          };
        }
        return { valid: true, severity: 'info' };
      }
      return {
        valid: false,
        message: `Field '${field.name}' value '${stringValue}' could not be reconciled. Best local match: '${bestMatch.canonicalValue}' (${(bestMatch.confidence * 100).toFixed(1)}%).`,
        severity: 'warning',
      };
    }

    return {
      valid: true,
      message: `Field '${field.name}': reconciliation requires async validation or a localDictionary config.`,
      severity: 'info',
    };
  }

  private findBestLocalMatch(
    value: string,
    candidates: string[]
  ): { canonicalValue: string; confidence: number } {
    let bestCanonical = candidates[0] ?? '';
    let bestConfidence = 0;
    const lowerValue = value.toLowerCase();

    for (const candidate of candidates) {
      const lowerCandidate = candidate.toLowerCase();
      if (lowerValue === lowerCandidate) {
        return { canonicalValue: candidate, confidence: 1.0 };
      }
      const similarity = this.jaroWinkler(lowerValue, lowerCandidate);
      if (similarity > bestConfidence) {
        bestConfidence = similarity;
        bestCanonical = candidate;
      }
    }

    return { canonicalValue: bestCanonical, confidence: bestConfidence };
  }

  private jaroWinkler(s1: string, s2: string): number {
    if (s1 === s2) return 1.0;
    const len1 = s1.length;
    const len2 = s2.length;
    if (len1 === 0 || len2 === 0) return 0.0;

    const matchDistance = Math.floor(Math.max(len1, len2) / 2) - 1;
    const s1Matches = new Array(len1).fill(false);
    const s2Matches = new Array(len2).fill(false);

    let matches = 0;
    let transpositions = 0;

    for (let i = 0; i < len1; i++) {
      const start = Math.max(0, i - matchDistance);
      const end = Math.min(i + matchDistance + 1, len2);
      for (let j = start; j < end; j++) {
        if (s2Matches[j] || s1[i] !== s2[j]) continue;
        s1Matches[i] = true;
        s2Matches[j] = true;
        matches++;
        break;
      }
    }

    if (matches === 0) return 0.0;

    let k = 0;
    for (let i = 0; i < len1; i++) {
      if (!s1Matches[i]) continue;
      while (!s2Matches[k]) k++;
      if (s1[i] !== s2[k]) transpositions++;
      k++;
    }

    const jaro =
      (matches / len1 + matches / len2 + (matches - transpositions / 2) / matches) / 3;

    let prefix = 0;
    for (let i = 0; i < Math.min(4, Math.min(len1, len2)); i++) {
      if (s1[i] === s2[i]) prefix++;
      else break;
    }

    return jaro + prefix * 0.1 * (1 - jaro);
  }

  private registerDefaultHandlers(): void {
    // Wikidata SPARQL handler (requires async invocation via validateAsync)
    this.knowledgeBaseHandlers.set('wikidata', async (value: string, options: Record<string, unknown>) => {
      const endpoint = (options.apiEndpoint as string) ?? 'https://www.wikidata.org/w/api.php';
      const url = `${endpoint}?action=wbsearchentities&search=${encodeURIComponent(value)}&language=en&format=json`;

      const response = await fetch(url);
      const data = await response.json() as { search?: Array<{ label: string; description?: string }> };

      if (!data.search || data.search.length === 0) return [];

      return data.search.slice(0, 5).map((item: { label: string; description?: string }) => ({
        canonicalValue: item.label,
        confidence: item.label.toLowerCase() === value.toLowerCase() ? 1.0 : 0.7,
        source: 'wikidata',
      }));
    });

    // Custom API handler
    this.knowledgeBaseHandlers.set('custom', async (value: string, options: Record<string, unknown>) => {
      const apiEndpoint = options.apiEndpoint as string | undefined;
      if (!apiEndpoint) throw new Error('Custom KB requires apiEndpoint configuration.');

      const response = await fetch(`${apiEndpoint}?q=${encodeURIComponent(value)}`);
      const data = await response.json() as { results?: Array<{ value: string; score: number }> };

      return (data.results ?? []).map((r: { value: string; score: number }) => ({
        canonicalValue: r.value,
        confidence: r.score,
        source: 'custom',
      }));
    });
  }

  appliesTo(field: FieldDef): boolean {
    return field.type.toLowerCase() === 'string';
  }

  dimension(): QualityDimension {
    return 'accuracy';
  }
}

export default ReconciliationQualityProvider;

File: kits/data-integration/providers/quality_rule/Reconciliation.swift
// Quality Rule Provider: Reconciliation Validation
// Matches field values against external knowledge bases for accuracy verification.
// Dimension: accuracy

import Foundation

public struct KBMatch {
    public let canonicalValue: String
    public let confidence: Double
    public let source: String
}

public final class ReconciliationQualityProvider {
    public typealias KBQueryHandler = (String, [String: Any]) async throws -> [KBMatch]
    private var knowledgeBaseHandlers: [String: KBQueryHandler] = [:]

    public init() {}

    /// Register a custom knowledge base query handler for async reconciliation.
    public func registerKnowledgeBase(name: String, handler: @escaping KBQueryHandler) {
        knowledgeBaseHandlers[name] = handler
    }

    /// Synchronous validate using local dictionary matching.
    public func validate(value: Any?, field: FieldDef, record: [String: Any], config: RuleConfig) -> RuleResult {
        guard let value = value, !(value is NSNull) else {
            return RuleResult(valid: true, message: nil, severity: "info")
        }

        let stringValue = String(describing: value)
        let matchThreshold = config.threshold ?? 0.8

        // Use local dictionary for synchronous reconciliation
        if let localValues = config.options?["localDictionary"] as? [String], !localValues.isEmpty {
            let bestMatch = findBestLocalMatch(value: stringValue, candidates: localValues)

            if bestMatch.confidence >= matchThreshold {
                if bestMatch.canonicalValue != stringValue {
                    return RuleResult(
                        valid: true,
                        message: "Field '\(field.name)': suggested canonical form is '\(bestMatch.canonicalValue)' (similarity: \(String(format: "%.1f", bestMatch.confidence * 100))%).",
                        severity: "info"
                    )
                }
                return RuleResult(valid: true, message: nil, severity: "info")
            }

            return RuleResult(
                valid: false,
                message: "Field '\(field.name)' value '\(stringValue)' could not be reconciled. Best match: '\(bestMatch.canonicalValue)' (\(String(format: "%.1f", bestMatch.confidence * 100))%).",
                severity: "warning"
            )
        }

        return RuleResult(
            valid: true,
            message: "Field '\(field.name)': reconciliation requires async validation or a localDictionary config.",
            severity: "info"
        )
    }

    /// Async validate using registered knowledge base handlers.
    public func validateAsync(value: Any?, field: FieldDef, record: [String: Any], config: RuleConfig) async -> RuleResult {
        guard let value = value, !(value is NSNull) else {
            return RuleResult(valid: true, message: nil, severity: "info")
        }

        let stringValue = String(describing: value)
        let matchThreshold = config.threshold ?? 0.8
        let knowledgeBase = config.options?["knowledgeBase"] as? String ?? "wikidata"

        guard let handler = knowledgeBaseHandlers[knowledgeBase] else {
            return RuleResult(
                valid: false,
                message: "Reconciliation rule for '\(field.name)': unknown knowledge base '\(knowledgeBase)'.",
                severity: "warning"
            )
        }

        do {
            let matches = try await handler(stringValue, config.options ?? [:])

            if matches.isEmpty {
                return RuleResult(
                    valid: false,
                    message: "Field '\(field.name)' value '\(stringValue)' not found in knowledge base '\(knowledgeBase)'.",
                    severity: "warning"
                )
            }

            let bestMatch = matches.max(by: { $0.confidence < $1.confidence })!

            if bestMatch.confidence >= matchThreshold {
                if bestMatch.canonicalValue != stringValue {
                    return RuleResult(
                        valid: true,
                        message: "Field '\(field.name)': canonical form is '\(bestMatch.canonicalValue)' (confidence: \(String(format: "%.1f", bestMatch.confidence * 100))%, source: \(bestMatch.source)).",
                        severity: "info"
                    )
                }
                return RuleResult(valid: true, message: nil, severity: "info")
            }

            return RuleResult(
                valid: false,
                message: "Field '\(field.name)' value '\(stringValue)' has low confidence match (\(String(format: "%.1f", bestMatch.confidence * 100))%) against '\(knowledgeBase)'.",
                severity: "warning"
            )
        } catch {
            return RuleResult(
                valid: false,
                message: "Reconciliation error for '\(field.name)': \(error.localizedDescription)",
                severity: "warning"
            )
        }
    }

    private func findBestLocalMatch(value: String, candidates: [String]) -> (canonicalValue: String, confidence: Double) {
        let lowerValue = value.lowercased()
        var bestCanonical = candidates.first ?? ""
        var bestConfidence = 0.0

        for candidate in candidates {
            if lowerValue == candidate.lowercased() {
                return (canonicalValue: candidate, confidence: 1.0)
            }
            let similarity = jaroWinkler(s1: lowerValue, s2: candidate.lowercased())
            if similarity > bestConfidence {
                bestConfidence = similarity
                bestCanonical = candidate
            }
        }

        return (canonicalValue: bestCanonical, confidence: bestConfidence)
    }

    private func jaroWinkler(s1: String, s2: String) -> Double {
        if s1 == s2 { return 1.0 }
        let c1 = Array(s1)
        let c2 = Array(s2)
        let len1 = c1.count
        let len2 = c2.count
        if len1 == 0 || len2 == 0 { return 0.0 }

        let matchDistance = max(len1, len2) / 2 - 1
        var s1Matches = Array(repeating: false, count: len1)
        var s2Matches = Array(repeating: false, count: len2)
        var matches = 0
        var transpositions = 0

        for i in 0..<len1 {
            let start = max(0, i - matchDistance)
            let end = min(i + matchDistance + 1, len2)
            for j in start..<end {
                guard !s2Matches[j], c1[i] == c2[j] else { continue }
                s1Matches[i] = true
                s2Matches[j] = true
                matches += 1
                break
            }
        }

        if matches == 0 { return 0.0 }

        var k = 0
        for i in 0..<len1 {
            guard s1Matches[i] else { continue }
            while !s2Matches[k] { k += 1 }
            if c1[i] != c2[k] { transpositions += 1 }
            k += 1
        }

        let m = Double(matches)
        let jaro = (m / Double(len1) + m / Double(len2) + (m - Double(transpositions) / 2.0) / m) / 3.0

        var prefix = 0
        for i in 0..<min(4, min(len1, len2)) {
            if c1[i] == c2[i] { prefix += 1 }
            else { break }
        }

        return jaro + Double(prefix) * 0.1 * (1.0 - jaro)
    }

    public func appliesTo(field: FieldDef) -> Bool {
        return field.type_.lowercased() == "string"
    }

    public func dimension() -> QualityDimension {
        return .accuracy
    }
}
